src/policy/fees_input.cpp

```diff
@@ -0,0 +1,222 @@
+};
+
+FeeEstInput::FeeEstInput(CBlockPolicyEstimator& estimator) : estimator(estimator) {}
+
+void FeeEstInput::processTx(const uint256& hash, unsigned int height, CAmount fee, size_t size, Block* block, bool valid)
+{
+    estimator.processTx(hash, height, fee, size, block ? block->est_block : nullptr, valid);
+    if (log) {
+        UniValue tx(UniValue::VOBJ);
+        tx.pushKV("hash", hash.ToString());
+        tx.pushKV("height", int(height));
+        tx.pushKV("fee", fee);
+        tx.pushKV("size", uint64_t(size));
+        if (block) {
+            block->json_txs.push_back(std::move(tx));
+        } else {
+            UniValue value(UniValue::VOBJ);
+            value.pushKV("tx", std::move(tx));
+            value.pushKV("valid", UniValue(valid));
+            value.pushKV("time", GetTime());
+            *log << value.write() << std::endl;
+        }
+    }
+}
+
+void FeeEstInput::processBlock(int height, const std::function<void(Block&)>& process_txs)
+{
+    Block block;
+    estimator.processBlock(height, [&](CBlockPolicyEstimator::Block& est_block) {
+        block.est_block = &est_block;
+        process_txs(block);
+    });
+    if (log) {
+        UniValue json(UniValue::VOBJ);
+        UniValue json_block(UniValue::VOBJ);
+        json_block.pushKV("height", height);
+        json.pushKV("block", json_block);
+        json.pushKV("txs", block.json_txs);
+        json.pushKV("time", GetTime());
+        *log << json.write() << std::endl;
+    }
+}
+
+void FeeEstInput::removeTx(const uint256& hash, bool in_block)
+{
+    estimator.removeTx(hash, in_block);
+
+    if (log) {
+        UniValue value(UniValue::VOBJ);
+        UniValue removeTx(UniValue::VOBJ);
+        removeTx.pushKV("hash", hash.ToString());
+        removeTx.pushKV("inBlock", UniValue(in_block));
+        value.pushKV("removeTx", removeTx);
+        value.pushKV("time", GetTime());
+        *log << value.write() << std::endl;
+    }
+}
+
+void FeeEstInput::flushUnconfirmed()
+{
+    estimator.FlushUnconfirmed();
+    if (log) {
+        UniValue value(UniValue::VOBJ);
+        UniValue flush(UniValue::VARR);
+        value.pushKV("flush", flush);
+        value.pushKV("time", GetTime());
+        *log << value.write() << std::endl;
+    }
+}
+
+bool FeeEstInput::writeData(const fs::path& filename)
+{
+    CAutoFile file(fsbridge::fopen(filename, "wb"), SER_DISK, CLIENT_VERSION);
+    if (file.IsNull()) {
+        LogPrintf("%s: Failed to write fee estimates to %s\n", __func__, filename.string());
+        return false;
+    }
+    estimator.Write(file);
+    return true;
+}
+
+bool FeeEstInput::readData(const fs::path& filename)
+{
+    CAutoFile file(fsbridge::fopen(filename, "rb"), SER_DISK, CLIENT_VERSION);
+    if (file.IsNull() || !estimator.Read(file)) {
+        return false;
+    }
+    if (log) {
+        UniValue value(UniValue::VOBJ);
+        std::ifstream file(filename.string(), std::ifstream::binary);
```

Comment:
The variable `file` shadows an already existing local variable. Please choose another name :-)